### PR TITLE
Work image zoom

### LINKED ISSUE #1125: Zoom in on an image
## Type
- ✨ Feature

## Requirements
We're pretty likely to want to offer deep zoom functionality beyond what can be done within native browser. What are our options for zooming in on IIIF images? 

Options in the world include: 
- http://iiif.io/apps-demos/
- OpenSeadragon
- https://viewdir.github.io/
- Universal Viewer
....and there are more I'm sure!

### PR DESCRIPTION
**Pull request by gestchild**
Still WIP as cross browser checking and want to add some more tracking, but be good to get some eyes on this.

References #1125

## Type
✨ Feature  

- [ ] Demoed to @Heesoomoon 

## Value
Allows users to see greater image detail

## Screenshot
screen shot 2017-10-12 at 10 35 16 2

### All
- [x] PR labelled and assigned
- [x] Any introduced code complexity has been flagged

### UI component only
- [x] A11y tested: `npm run test:accessibility <URL>`
- [-] Structured data tested: https://search.google.com/structured-data/testing-tool 
- [ ] Browser tested: `npm run test:browsers <URL>`
- [-] Works without JS in the client
- [ ] Relevant tracking/monitoring has been considered

### CODE REVIEW

File: client/js/app.js [hunk 1]
```diff
@@ -58,19 +58,17 @@ const init = () => {
   const backToTopEl = document.querySelector('.js-back-to-top');
   const toggleShowHideEls = document.querySelectorAll('.js-show-hide');
   const scrollToInfoEls = document.querySelectorAll('.js-scroll-to-info');
-  const workMediaEls = document.querySelectorAll('.js-work-media');
   const copyUrlEls = document.querySelectorAll('.js-copy-url');
   const searchBoxEls = document.querySelectorAll('.js-search-box');
   const iframeContainerEls = document.querySelectorAll('.js-iframe-container');
+  const fullscreenViewers = document.querySelectorAll('.js-image-viewer-fullscreen');
```
Comment (jamesgorrie) [inline client/js/app.js]: Just wondering, what's better, `fullscreenViewer`, or `viewerFullscreen`?
We seem to have both?

File: client/js/components/image-viewer.js [hunk 2]
```diff
@@ -0,0 +1,72 @@
+import { hasFullscreen, enterFullscreen, exitFullscreen } from '../util';
+import fastdom from '../utils/fastdom-promise';
+import { trackGaEvent } from '../tracking';
+import OpenSeadragon from 'openseadragon';
+
+function setupViewer(imageInfoSrc, viewer, viewerId) {
+  if (viewer.querySelector('.openseadragon-container')) return;
+  window.fetch(imageInfoSrc)
+  .then(function(response) {
```
Comment (jamesgorrie) [inline client/js/components/image-viewer.js]: You could use `res => res.json` for brevity.

File: client/js/components/image-viewer.js [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    fastdom.mutate(() => {
+      viewer.style.display = 'block';
+    });
+
+    image.addEventListener('dblclick', (e) => {
+      setupViewer(imageInfoSrc, viewer, viewerId);
+      enterFullscreen(viewerContent);
+      trackGaEvent({
+        category: 'component',
+        action: 'work-enter-fullscreen-image:dblclick',
```
Comment (jamesgorrie) [inline client/js/components/image-viewer.js]: I can't say this action makes all that much sense to me, but I'm not that close to the interaction.
Is it when you enter fullscreen via dblclick?

File: client/js/components/image-viewer.js [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        label: `id:${window.location.pathname.match(/\/works\/(.+)/)[1]}, title:${document.title}`
```
Comment (jamesgorrie) [inline client/js/components/image-viewer.js]: Why are we storing the title in the analytics?
Could we not get this info from the component rather?

File: client/scss/components/_fullscreen_viewer.scss [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    display: none;
+
+    @include respond-to('medium') {
+      display: inline-block;
+    }
+  }
+}
+
+.fullscreen-viewer__image {
+  background: #1d1d1d;
```
Comment (jamesgorrie) [inline client/scss/components/_fullscreen_viewer.scss]: ❓

File: client/scss/components/_image_viewer.scss [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+  width: 100%;
+  height: calc(100vh - 3em);
+}
+
+.image-viewer-fullscreen__launch-button {
+  position: absolute;
+  bottom: $v-spacing-unit*4;
+  left: 50%;
+  transform: translateX(-50%);
+}
```
Comment (jamesgorrie) [inline client/scss/components/_image_viewer.scss]: Do we ever need the fullscreen not with the image viewer?
What's the reason for the decoupling?

File: client/scss/components/_scroll_to_info.scss [hunk 7]
```diff
@@ -1,3 +1,14 @@
 .scroll-to-info {
+  display: none;
   bottom: 4 * $spacing-unit;
+  outline: none;
```
Comment (jamesgorrie) [inline client/scss/components/_scroll_to_info.scss]: Accessibility?

File: server/filters/convert-image-uri.js [hunk 8]
```diff
@@ -57,7 +57,15 @@ function convertPathToIiifUri(originalUriPath, iiifRoot, size) {
   return `${iiifRoot}${originalUriPath}/full/${size}${isFullSize ? '' : ','}/0/default.${format}`;
 }
 
-export default function convertImageUri(originalUri, requiredSize, useIiifOrigin) {
+export function convertIiifUriToInfoUri(originalUriPath) {
+  if (originalUriPath.startsWith('https://iiif')) {
```
Comment (jamesgorrie) [inline server/filters/convert-image-uri.js]: Do we (polyfil / are happy not to) this?

File: server/filters/generate-id.js [hunk 9]
```diff
@@ -0,0 +1,5 @@
+import shortid from 'shortid';
+
+export default function generateId(blank) {
+  return shortid.generate();
+};
```
Comment (jamesgorrie) [inline server/filters/generate-id.js]: Why can we not use the image id?

File: server/views/components/button-button/button-button.njk [hunk 10]
```diff
@@ -1,6 +1,6 @@
-<button class="btn {{ {s:'HNM5'} | fontClasses }} {{ modifierClasses }}" {{ attributes }}>
+<button {{ 'id='+ buttonId if buttonId }} class="btn {{ modifierClasses }} {{ {s:'HNM5'} | fontClasses }}" {{ attributes }}{% if eventTracking %} data-track-event="{{eventTracking}}"{% endif %}>
```
Comment (jamesgorrie) [inline server/views/components/button-button/button-button.njk]: buttonId won't be quoted.

File: server/views/components/image-viewer/image-viewer.njk [hunk 11]
```diff
@@ -0,0 +1,24 @@
+{% set viewerId = '' | generateId %}
+{% set buttonFontClasses =  {s:'HNM5'} | fontClasses %}
+{% set buttonSpacingClasses =  {s: 1} | spacingClasses({margin: ['top', 'right', 'bottom', 'left']}) %}
+{% set commonBtnTracking = '"category": "component", "label": "id:' + data.id + ', title:' + data.trackTitle + '"' %}
+<div class="js-image-viewer-fullscreen image-viewer-fullscreen">
+
+  {% component 'button-button', {text: 'View larger image', icon: 'actions/expand', modifierClasses: buttonFontClasses + ' ' + buttonSpacingClasses + ' image-viewer-fullscreen__launch-button js-enter-fullscreen', eventTracking: '{' + commonBtnTracking + ', "action": "work-enter-fullscreen-button:click"}' } %}
```
Comment (jamesgorrie) [inline server/views/components/image-viewer/image-viewer.njk]: could we array join here to make it a little neater?

File: client/js/components/image-viewer.js (hunk 3 above)
Comment (gestchild) [reply client/js/components/image-viewer.js]: exactly that yep

File: client/scss/components/_scroll_to_info.scss (hunk 7 above)
Comment (gestchild) [reply client/scss/components/_scroll_to_info.scss]: we put our own focus style on it

File: server/filters/convert-image-uri.js (hunk 8 above)
Comment (gestchild) [reply server/filters/convert-image-uri.js]: This isn't browser code

File: server/filters/generate-id.js (hunk 9 above)
Comment (gestchild) [reply server/filters/generate-id.js]: er, yeah we can... not sure why I did this now

File: client/js/app.js [hunk 12]
```diff
@@ -79,6 +77,8 @@ const init = () => {
 
   nodeList(scrollToInfoEls).forEach(scrollToInfo);
 
+  nodeList(viewersFullscreen).forEach((viewer) => createImageViewer(viewer));
```
Comment (davidpmccormick) [inline client/js/app.js]: `nodeList(viewersFullscreen).forEach(createImageViewer)` is more hip!

File: client/js/components/image-viewer.js [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+        'width': response.width,
+        'profile': [ 'http://iiif.io/api/image/2/level2.json' ],
+        'protocol': 'http://iiif.io/api/image',
+        'tiles': [{
+          'scaleFactors': [ 1, 2, 4, 8, 16, 32 ],
+          'width': 400
+        }]
+      }]
+    });
+  }).catch(err => { throw err; });
```
Comment (davidpmccormick) [inline client/js/components/image-viewer.js]: do we need to do something more elegant with the error?

File: client/scss/components/_divider.scss [hunk 14]
```diff
@@ -13,6 +13,10 @@
   background-color: color('pumice');
 }
 
+.divider--charcoal {
```
Comment (davidpmccormick) [inline client/scss/components/_divider.scss]: I guess all of these modifiers aren't necessary anymore – we could use e.g. `bg-charcoal`.

**Pull request merged**